### DISCUSSION TITLE
clone syscall

### DISCUSSION THREAD

**Discussion by hel-asli** [Mouad-kimdil/execution#4] [General]
how fork use clone system calls to create a new child process ?

**Comment by Pilot-61** [comment]
In Bash, when you run commands or scripts, the shell itself uses the fork system call to create new processes. This is a fundamental part of how Unix-like operating systems manage processes.

Here’s a brief explanation of how fork and clone work in this context:

fork System Call
Purpose: fork is used to create a new process by duplicating the calling process. The new process is known as the child process, and it is an exact copy of the parent process (the process that called fork).
Behavior: After fork, both the parent and the child processes will continue executing the code following the fork call. They have separate memory spaces but share some resources.
clone System Call
Purpose: clone is a more advanced system call that is used to create new processes or threads. It provides more control over what is shared between the parent and child processes compared to fork.
Behavior: With clone, you can specify which parts of the process memory should be shared (e.g., file descriptors, memory space). It's commonly used in creating threads or lightweight processes.
In Bash
When you run a command in Bash, the shell performs the following steps:

Fork: Bash uses fork to create a new process. This new process is a copy of the Bash shell process.
Execute: In the child process, Bash then uses exec to replace the process image with the program you want to run (e.g., a command or script).
Wait: The parent process (the original Bash shell) waits for the child process to finish executing the command.
In this example:

The ls command is executed in a new child process created by fork and run in the background (&).
The echo command continues in the parent shell process.
Summary
fork: Creates a new process by duplicating the current process. This is used by Bash to start new commands or scripts.
clone: More advanced, used for creating threads or processes with more control over shared resources.
In the context of running commands in Bash, the fork system call is typically used to create a new process for executing the command, and exec is used to replace the process image with the command you want to run.

**Comment by Pilot-61** [comment]
> how fork use clone system calls to create a new child process ?
>fork and clone System Calls
In Unix-like operating systems, both fork and clone system calls are used for process creation, but they serve slightly different purposes and offer different levels of control.
